sympy/solvers/inequalities.py

```diff
@@ -982,3 +986,482 @@ def reduce_inequalities(inequalities, symbols=[]):
 
     # restore original symbols and return
     return rv.xreplace({v: k for k, v in recast.items()})
+
+
+class UnboundedLinearProgrammingError(Exception):
+    """
+    A linear programing problem is said to be unbounded if its objective
+    function can assume arbitrarily large values.
+
+    Example
+    =======
+
+    Suppose you want to maximize
+        2x
+    subject to
+        x >= 0
+
+    There's no upper limit that 2x can take.
+    """
+    pass
+
+
+class InfeasibleLinearProgrammingError(Exception):
+    """
+    A linear programing problem is considered infeasible if its constraint set
+    is empty. That is, if the set of all vectors satisfying the contraints is
+    empty, then the problem is infeasible.
+
+    Example
+    =======
+
+    Suppose you want to maximize
+        x
+    subject to
+        x >= 10
+        x <= 9
+
+    It's not possible for x to satisfy the given contraints.
+    """
+    pass
+
+
+def _pivot(M, i, j):
+    """
+    There are four rules for the pivot operation:
+    - The pivot quantity goes into its reciprocal
+    - Entries in the same row as the pivot are divided by the pivot
+    - Entries in the same column as the pivot are divided by the pivot and
+    changed in sign.
+    - The remaining entries are reduced in value by the following: the product
+    of the corresponding entries in the same row and column as themselves and
+    the pivot, divided by the pivot.
+
+    Example
+    =======
+
+    >>> from sympy.matrices.dense import Matrix
+    >>> from sympy.solvers.inequalities import _pivot
+    >>> m = Matrix([[3,1,5],
+    ...             [2,-3,1],
+    ...             [0,3,1]])
+    >>> _pivot(m, 1,0) # pivot around 2, which is in row 1 and col 0
+    Matrix([
+    [-3/2, 11/2, 7/2],
+    [ 1/2, -3/2, 1/2],
+    [   0,    3,   1]])
+    """
+    Mi, Mj, Mij = M[i,:], M[:,j], M[i,j]
+    if Mij == 0:
+        raise ZeroDivisionError("Tried to pivot about zero-valued entry.")
+    MM = M - Mj * (Mi / Mij)
+    MM[i,:] = Mi / Mij
+    MM[:,j] = -Mj / Mij
+    MM[i,j] = 1 / Mij
+    return MM
+
+
+def _choose_pivot_row(A, B, candidate_rows, pivot_col, S):
+    # Choose row with smallest ratio
+    min_ratio = float('inf')
```

Comment:
I see that `oo == float('inf') == Float('inf')` -- is there any reason to choose the latter instead of the first?

Reply:
It would be better to avoid using any infinity. Just initialise this to the first item.

I have actually measured situations in which the use of `oo` causes slowdowns because `dup_degree` returns `-oo`:
https://github.com/sympy/sympy/blob/50eb4007d3305f115cf30217a5235b5ff056626d/sympy/polys/densebasic.py#L153
The problem is that evaluating `-oo < 0` is much slower than evaluating `float('-inf') < 0`. In the case of `dup_degree` we know that the comparison will always be with an `int` so we should just use `float('-inf')` if really necessary. A symbolic object like `oo` should not be used in low-level routines like the internals of `Poly`. Better would just be a `dup_is_zero` function to avoid depending on infinities at all.
```python
In [1]: ninf1 = -oo

In [2]: ninf2 = float('-inf')

In [3]: %timeit ninf1 < 0
33.3 µs ± 964 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [4]: %timeit ninf2 < 0
43.1 ns ± 1.04 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
That's a 1000x speed difference: nanoseconds vs microseconds.